File: backend-php/app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Casts\Attribute;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Product extends Model
{
    use HasFactory;

    protected $fillable = [
        'name',
        'description',
        'price',
        'quantity',
        'low_stock_threshold',
        'category_id',
        'specifications',
        'attributes',
        'is_active',
    ];

    protected $casts = [
        'price' => 'decimal:2',
        'quantity' => 'integer',
        'low_stock_threshold' => 'integer',
        'specifications' => 'array',
        'attributes' => 'array',
        'is_active' => 'boolean',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    protected $appends = ['is_low_stock', 'primary_image_url'];

    public function category(): BelongsTo
    {
        return $this->belongsTo(Category::class);
    }

    public function images(): HasMany
    {
        return $this->hasMany(ProductImage::class);
    }

    public function comments(): HasMany
    {
        return $this->hasMany(Comment::class);
    }

    public function inventoryLogs(): HasMany
    {
        return $this->hasMany(InventoryLog::class);
    }

    protected function isLowStock(): Attribute
    {
        return Attribute::get(fn () => $this->quantity <= $this->low_stock_threshold);
    }

    protected function primaryImageUrl(): Attribute
    {
        return Attribute::get(function () {
            $images = $this->relationLoaded('images') ? $this->images : $this->images()->get();
            $primary = $images->firstWhere('is_primary', true) ?: $images->first();
            return $primary ? $primary->image_url : 'https://cdn.image.com/example.jpg';
        });
    }
}

File: backend-fastapi/app/schemas/product.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ── Request Schemas ────────────────────────────────────────────────────

class ProductRequest(BaseModel):
    name: str = Field(..., max_length=200)
    description: Optional[str] = None
    price: float = Field(..., gt=0)
    quantity: Optional[int] = Field(0, ge=0)
    low_stock_threshold: Optional[int] = Field(10, ge=0)
    category_id: int
    specifications: Optional[dict] = None
    attributes: Optional[dict] = None


class ProductWithImageUrlsRequest(ProductRequest):
    image_urls: Optional[list[str]] = None


# ── Response Schemas ───────────────────────────────────────────────────

class ProductImageResponse(BaseModel):
    id: int
    image_url: str
    is_primary: bool

    model_config = ConfigDict(from_attributes=True)


class ProductResponse(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    price: float
    quantity: int
    low_stock_threshold: int
    category_id: int
    category_name: Optional[str] = None
    specifications: Optional[dict] = None
    attributes: Optional[dict] = None
    images: list[ProductImageResponse] = []
    primary_image_url: Optional[str] = None
    is_active: bool
    is_low_stock: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)
